File: src/main/java/com/felipe/booking/app/handler/BookingHandler.java
package com.felipe.booking.app.handler;

import com.felipe.booking.app.dto.request.BookingRequestDTO;
import com.felipe.booking.app.dto.response.BookingResponseDTO;
import com.felipe.booking.app.dto.response.RoomAvailabilityDTO;
import com.felipe.booking.domain.usecase.DeleteBookingUseCase;
import com.felipe.booking.domain.usecase.FindBookingUseCase;
import com.felipe.booking.domain.usecase.SaveBookingUseCase;
import com.felipe.booking.domain.usecase.GetRoomAvailabilityUseCase;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;

import java.net.URI;

@Component
public class BookingHandler {

    @Autowired
    private SaveBookingUseCase saveBookingUseCase;

    @Autowired
    private DeleteBookingUseCase deleteBookingUseCase;

    @Autowired
    private FindBookingUseCase findBookingUseCase;

    @Autowired
    private GetRoomAvailabilityUseCase getRoomAvailabilityUseCase;

    public Mono<ServerResponse> createBooking(ServerRequest serverRequest) {
        return serverRequest
                .bodyToMono(BookingRequestDTO.class)
                .map(it -> it.toDomain())
                .flatMap(it -> saveBookingUseCase.execute(it))
                .flatMap(it -> ServerResponse.created(URI.create(it.getId())).bodyValue(BookingResponseDTO.of(it)));
    }

    public Mono<ServerResponse> saveBooking(ServerRequest serverRequest) {
        return serverRequest
                .bodyToMono(BookingRequestDTO.class)
                .map(it -> it.toDomain(serverRequest.pathVariable("bookingId")))
                .flatMap(it -> saveBookingUseCase.execute(it))
                .flatMap(it -> ServerResponse.status(HttpStatus.ACCEPTED).bodyValue(BookingResponseDTO.of(it)));
    }

    public Mono<ServerResponse> deleteBooking(ServerRequest serverRequest) {
        return Mono.just(serverRequest)
                .flatMap(it -> deleteBookingUseCase.delete(serverRequest.pathVariable("bookingId")))
                .flatMap(it -> ServerResponse.noContent().build());
    }

    public Mono<ServerResponse> findBooking(ServerRequest serverRequest) {
        return Mono.just(serverRequest)
                .flatMap(it -> findBookingUseCase.find(serverRequest.pathVariable("bookingId")))
                .flatMap(it -> ServerResponse.ok().bodyValue(BookingResponseDTO.of(it)));
    }

    public Mono<ServerResponse> getRoomAvailability(ServerRequest serverRequest) {
        return Mono.just(serverRequest)
                .flatMap(it -> getRoomAvailabilityUseCase.execute())
                .flatMap(it -> ServerResponse.ok().bodyValue(RoomAvailabilityDTO.of(it)));
    }

}

File: src/main/kotlin/com/felipe/booking/alten/app/handler/BookingHandler.kt
package com.felipe.booking.alten.app.handler

import com.felipe.booking.alten.app.dto.BookingDTO
import com.felipe.booking.alten.app.dto.RoomAvailabilityDTO
import com.felipe.booking.alten.domain.usecase.CreateBookingUseCase
import com.felipe.booking.alten.domain.usecase.GetRoomAvailabilityUseCase
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.server.ServerRequest
import org.springframework.web.reactive.function.server.ServerResponse
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.toMono

@Component
class BookingHandler(
    private val createBookingUseCase: CreateBookingUseCase,
    private val getRoomAvailabilityUseCase: GetRoomAvailabilityUseCase
) {

    fun createBooking(serverRequest: ServerRequest): Mono<ServerResponse> =
        serverRequest
            .bodyToMono(BookingDTO::class.java)
            .map { it.toDomain() }
            .flatMap { createBookingUseCase.execute(it) }
            .map { BookingDTO.of(it) }
            .flatMap { ServerResponse.ok().bodyValue(it) }

    fun getRoomAvailability(serverRequest: ServerRequest): Mono<ServerResponse> =
        serverRequest
            .toMono()
            .flatMap { getRoomAvailabilityUseCase.execute() }
            .map { RoomAvailabilityDTO.of(it) }
            .flatMap { ServerResponse.ok().bodyValue(it) }
}